### ISSUE TITLE
mesolve with LowRank method evaluates a full dense density matrix in save_extra

### ISSUE THREAD

**Issue by norareinic** [dynamiqs/dynamiqs#1132]
@p-pietro @gautierronan 

Right now, when running mesolve with `method=LowRank()` and measuring a quantity via a custom funtion in save_extra, the custom function operates on a full dense density matrix (dim = n x n, where n is the total Hilbert space dimension). This means that the simulation is going to be limited by the memory required to store a full state, which is exactly what we wanted to avoid with introducing the low-rank method.

**Proposal:**
I propose that, instead, the custom functions operate on a low-rank density matrix, whose dimension is n x K, K<<n being the chosen rank.

**Arguments:** 
- Unless `save_states=True` (which again eats up a lot of memory, so we want to avoid it), I see no other way of measuring useful stuff like the Wigner function
- Measuring expectation values via exp_ops currently evaluates the full operators. This is a bug that should ideally be resolved (I will open another issue for this), but before it gets resolved we could measure expectation values via custom functions if neccessary
- Altogether, without this feature there is no much advantage in using LowRank with respect to the standard methods, and benchmarking results imply the same. Based on benchmarking, the gain in available Hilbert space sizes that we get if we avoid evaluating a full state could be huge

**Downsides of a proposal:**
- User has to be aware that the custom functions need to be defined for low-rank form and not the full state. But, if one wants to return to the full form, it's a one-line change in the custom function, we should just make sure to document it properly

**Existing code** ([dynamiqs/integrators/core/low_rank_integrator.py](https://github.com/dynamiqs/dynamiqs/blob/main/dynamiqs/integrators/core/low_rank_integrator.py) line `301-320`):
```
def save(self, y: PyTree) -> SolveSaved:
        m = y
        m = normalize_m(m)

        msave = None
        if self.options.save_states:
            msave = asqarray(m, dims=self.dims)

        extra = None
        if self.options.save_extra:
            rho = asqarray(rho_from_m(m), dims=self.dims)
            extra = self.options.save_extra(rho)

        if self.Es_jax is not None:
            Esave = jnp.stack([expval_from_m(m, E) for E in self.Es_jax])
        else:
            Esave = None

        return SolveSaved(ysave=msave, extra=extra, Esave=Esave)
```

**Change:**
instead of 
```
     if self.options.save_extra:
            rho = asqarray(rho_from_m(m), dims=self.dims)
            extra = self.options.save_extra(rho)
```
call a function on m
```
     if self.options.save_extra:
            extra = self.options.save_extra(m)
```

**Comment by p-pietro** [comment]
What would you think of solving this with an additional `save_extra` field inside method? In this way we would preserve the behaviour of the normal `save_extra` for all `mesolve` methods, while also exposing a method specific capability. It could look like this:

```
method = dq.method.LowRank(
    rank=4,
    key=jax.random.PRNGKey(0),
    save_extra=lambda m: jnp.linalg.norm(m, axis=-2),  # m has shape (n, rank)
)

result = dq.mesolve(H, jump_ops, rho0, tsave, method=method)
result.extra
```

Instead of:

```
result = dq.mesolve(
    H, jump_ops, rho0, tsave,
    method=method,
    save_extra=lambda rho: dq.expect(op, rho),  # rho is (n, n)
)

result.extra
```

**Comment by norareinic** [comment]
Sounds good!

**Comment by gautierronan** [comment]
Hi @norareinic, thanks a lot for raising this. I agree it's a bit annoying, but I'm not entirely convinced by either of the two solutions you and @p-pietro propose.

Regarding solution 1 (changing the signature of `save_extra` from `rho: QArray -> PyTree` to `m: QArray -> PyTree`):
- the main issue is that we're changing the behavior of an argument (`save_extra`) depending on the value of a completely separate argument (`method`). That is really not obvious from the user standpoint, and subject to a lot of frustration. For instance, if you write a simulation with `save_extra` that uses `method.Tsit5` and want to try out `method.LowRank`, then you'll face a silent failure.

Regarding solution 2 (passing `save_extra` directly inside `method.LowRank`):
- how do you deal with a user passing `save_extra` both inside the method and solver? What do you return inside `result.extra`? I think the answer is not obvious, and it's thus probably not a good API even if it's well documented.

Maybe a third solution that answers both of these concerns is to keep `save_extra` inside the solver arguments, but to add a boolean argument to `method.LowRank` that decides whether the input of `save_extra` should be the full matrix `rho`, or the low-rank one `m`. Such an argument should be `False` by default (take `rho` as input) such that the user makes a conscious decision when setting to `True`. What do you think?

**Comment by p-pietro** [comment]
Sounds good, I will submit a PR to this effect.

**Comment by norareinic** [comment]
Great, thanks a lot!
